dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.cpp

```diff
@@ -125,18 +130,35 @@ ColumnFileTiny::getReader(const DMContext & /*context*/, const StorageSnapshotPt
 
 void ColumnFileTiny::serializeMetadata(WriteBuffer & buf, bool save_schema) const
 {
-    serializeSchema(buf, save_schema ? schema : BlockPtr{});
+    serializeSchema(buf, save_schema ? schema->getSchema() : Block{});
 
     writeIntBinary(data_page_id, buf);
     writeIntBinary(rows, buf);
     writeIntBinary(bytes, buf);
 }
 
-std::tuple<ColumnFilePersistedPtr, BlockPtr> ColumnFileTiny::deserializeMetadata(ReadBuffer & buf, const BlockPtr & last_schema)
+ColumnFilePersistedPtr ColumnFileTiny::deserializeMetadata(DMContext & context, ReadBuffer & buf, ColumnFileSchemaPtr & last_schema)
 {
-    auto schema = deserializeSchema(buf);
-    if (!schema)
+    auto schema_block = deserializeSchema(buf);
+    std::shared_ptr<ColumnFileSchema> schema;
+
+    if (!schema_block)
+    {
         schema = last_schema;
+    }
+    else
+    {
+        auto new_digest = hashSchema(*schema_block);
+        schema = context.db_context.getSharedBlockSchemas()->find(new_digest);
+
+        if (schema == nullptr)
+        {
+            schema = std::make_shared<ColumnFileSchema>(*schema_block);
+            context.db_context.getSharedBlockSchemas()->insert(new_digest, schema);
+        }
+        last_schema = schema;
```

Comment:
Why do we need to assign last_schema?

Reply:
When we write schema into disk, we will only write the schema(when it changes). 
So for the continuous columnFileTiny with the same schema, we will only write the schema in first columnFileTiny.
Thus, when we do restore to read schema, we can only read the schema in limited columnFileTiny.
For the ColumnFileTiny we don't read the schema from disk, we just use the latest previous schema we read.
Thus, we need to assign last_schema here.

Reply:
Ah, I get it, so it is an output. Prefer to be a return value (just like previous code).